### PR TITLE
Ampers - Lily Sky

### PR DESCRIPTION
**Pull request by lillers1122**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I had to decide how much to include in my request_trip method in the dispatcher file. I decided to break it into two main methods (pick_driver and request_trip) so that the responsibilities were separated.
Describe any examples of composition that you encountered in this project, if any | Has-a: a passenger has-a trip.
Describe a _nominal test_ that you wrote for this assignment. | In the trip_dispatcher_spec I wrote a nominal test that when the dispatcher requests a trip it has to be an instance of a RideShare::Trip.
Describe an _edge case test_ that you wrote for this assignment. | In driver_spec I wrote an edge case test that a driver with a negative cost for a trip won't include that cost in their total revenue.
Describe a concept that you gained more clarity on as you worked on this assignment. | I gained more experience with when and how to call methods.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -1,9 +1,11 @@
 require 'csv'
 require_relative 'trip'
+require 'awesome_print'
 
 module RideShare
   class Driver
-    attr_reader :id, :name, :vehicle_id, :status, :trips
+    attr_reader :id, :name, :vehicle_id
+    attr_accessor :status, :trips
```
Comment (CheezItMan) [inline lib/driver.rb:8]: `trips` should never be an accessor.  You can still push new trips into the array with it as a reader.

File: lib/driver.rb [hunk 2]
```diff
@@ -17,31 +19,64 @@ def initialize(input)
...
+        elsif trip.cost != nil
+          total_revenue += (trip.cost - 1.65) * 0.8
+        end
+      end
+      return total_revenue.round(2)
+    end
+
+    def rate
+      driving_time = 0
+      @trips.each do |trip|
```
Comment (CheezItMan) [inline lib/driver.rb:67]: Making driving time it's own method would be a good choice. This works however.

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -86,16 +88,68 @@ def load_trips
...
+      drivers_latest_trips = []
+
+      available_drivers.each do |driver|
+        if driver.trips.max_by {|drive| drive.end_time} != nil
+          last_trip = driver.trips.max_by {|drive| drive.end_time}
+          drivers_latest_trips << {driver_id: driver.id, end_time: last_trip.end_time}
+        end
+      end
+
+      if drivers_latest_trips.length == 0 && newbie_drivers.length == 0
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:107]: Good!

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -72,9 +81,100 @@
...
 
-    it "returns zero if no trips" do
+    it "returns rating of nil if there are no trips" do
       driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
-      driver.average_rating.must_equal 0
+      driver.average_rating.must_equal nil
+    end
+  end
+
+  describe "total_revenue" do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:90]: What about a driver with no trips?

File: specs/trip_dispatch_spec.rb [hunk 5]
```diff
@@ -82,11 +83,109 @@
...
+      @dispatcher = RideShare::TripDispatcher.new
+      @dispatcher.request_trip(1).driver.id.must_equal 100
+      @dispatcher.request_trip(1).driver.id.must_equal 14
+      @dispatcher.request_trip(2).driver.id.must_equal 27
+      @dispatcher.request_trip(3).driver.id.must_equal 6
+      @dispatcher.request_trip(4).driver.id.must_equal 87
+      @dispatcher.request_trip(5).driver.id.must_equal 75
+    end
+
+    it "accurately raises an error when no drivers are available" do
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:114]: 👍

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits and overall good commit messages.
Answer comprehension questions | Check, Well done
**Wave 1** |
`Trip` has a helper method to calculate duration | Check
`Passenger` has a method to calculate total cost of all trips | Check
`Driver` has a method to calculate total revenue | Check
Complete tests | Very nicely done!
**Wave 2** |
`TripDispatcher` has a new method to create trips |  Check
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | Check
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips |  Check
Complete tests | Check
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | Well done, good helper methods
Appropriate use of Ruby's `Time` | Check
Appropriate helper methods were made to help with complex logic | Check
Complete tests | Well done!
**Overall** | Awesomework, Great job in testing and the complex logic of finding a driver, even accounting for drivers who've never had trips.  You really showed some mastery here.  Well done.
